client/clientutil/modelinfo.go

```diff
@@ -0,0 +1,461 @@
+			fmt.Fprintf(w, "%s:\t\n", headerName)
+			for _, elem := range headerIfaceList {
+				headerStringElem, ok := elem.(string)
+				if !ok {
+					return invalidTypeErr
+				}
+				// note we don't wrap these, since for now this is
+				// specifically just required-snaps and so all of these
+				// will be snap names which are required to be short
+				fmt.Fprintf(w, "  - %s\n", headerStringElem)
+			}
+
+		// timestamp needs to be formatted with fmtTime from the timeMixin
+		case "timestamp":
+			timestamp, ok := headerValue.(string)
+			if !ok {
+				return invalidTypeErr
+			}
+
+			// parse the time string as RFC3339, which is what the format is
+			// always in for assertions
+			t, err := time.Parse(time.RFC3339, timestamp)
+			if err != nil {
+				return err
+			}
+			fmt.Fprintf(w, "timestamp:\t%s\n", fmtTime(t, options.AbsTime))
+
+		// long string key we don't want to rewrap but can safely handle
+		// on "reasonable" width terminals
+		case "device-key-sha3-384":
+			// also flush the writer before continuing so the previous keys
+			// don't try to align with this key
+			w.Flush()
+			headerString, ok := headerValue.(string)
+			if !ok {
+				return invalidTypeErr
+			}
+
+			switch {
+			case options.TermWidth > 86:
+				fmt.Fprintf(w, "device-key-sha3-384: %s\n", headerString)
+			case options.TermWidth <= 86 && options.TermWidth > 66:
+				fmt.Fprintln(w, "device-key-sha3-384: |")
+				strutil.WordWrapPadded(w, []rune(headerString), "  ", options.TermWidth)
+			}
+		case "snaps":
+			// also flush the writer before continuing so the previous keys
+			// don't try to align with this key
+			w.Flush()
+			snapsHeader, ok := headerValue.([]interface{})
+			if !ok {
+				return invalidTypeErr
+			}
+			if len(snapsHeader) == 0 {
+				// unexpected why this is an empty list, but just ignore for
+				// now
+				continue
+			}
+			fmt.Fprintf(w, "snaps:\n")
+			for _, sn := range snapsHeader {
+				snMap, ok := sn.(map[string]interface{})
+				if !ok {
+					return invalidTypeErr
+				}
+				// iterate over all keys in the map in a stable, visually
+				// appealing ordering
+				// first do snap name, which will always be present since we
+				// parsed a valid assertion
+				name := snMap["name"].(string)
+				fmt.Fprintf(w, "  - name:\t%s\n", name)
+
+				// the rest of these may be absent, but they are all still
+				// simple strings
+				for _, snKey := range []string{"id", "type", "default-channel", "presence"} {
+					snValue, ok := snMap[snKey]
+					if !ok {
+						continue
+					}
+					snStrValue, ok := snValue.(string)
+					if !ok {
+						return invalidTypeErr
+					}
+					if snStrValue != "" {
+						fmt.Fprintf(w, "    %s:\t%s\n", snKey, snStrValue)
+					}
+				}
+
+				// finally handle "modes" which is a list
+				modes, ok := snMap["modes"]
+				if !ok {
+					continue
+				}
+				modesSlice, ok := modes.([]interface{})
+				if !ok {
+					return invalidTypeErr
+				}
+				if len(modesSlice) == 0 {
+					continue
```

Comment:
I'd rather not use `continue` here, because changing this code to add one more key will then require a refactoring. Either invert the conditions to only print the modes if they exist (but the nesting level will explode, it's already too deep here) or, and I'd vote for this, create a separate function to collect the mode list and use it here:
```golang
modes, err := snapModes(snMap)
if err != nil {
    return err
}
if len(modes) > 0 {
    modesSliceYamlStr := "[" + strings.Join(modes, ", ") + "]"
    fmt.Fprintf(w, "    modes:\t%s\n", modesSliceYamlStr)
}
```
Or even pass the writer `w` to the function.

Reply:
I have done something more alike to this now, I hope you agree with the new approach.

Reply:
Thanks!